clippy_lints/src/trivially_copy_pass_by_ref.rs

```diff
@@ -0,0 +1,148 @@
+        }
+
+        match kind {
+            FnKind::ItemFn(.., abi, _, attrs) => {
+                if abi != Abi::Rust {
+                    return;
+                }
+                for a in attrs {
+                    if a.meta_item_list().is_some() && a.name() == "proc_macro_derive" {
+                        return;
+                    }
+                }
+            },
+            FnKind::Method(..) => (),
+            _ => return,
+        }
+
+        // Exclude non-inherent impls
+        if let Some(NodeItem(item)) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(node_id)) {
+            if matches!(item.node, ItemImpl(_, _, _, _, Some(_), _, _) |
+                ItemTrait(..))
+            {
+                return;
+            }
+        }
+
+        let fn_def_id = cx.tcx.hir.local_def_id(node_id);
+
+        let fn_sig = cx.tcx.fn_sig(fn_def_id);
+        let fn_sig = cx.tcx.erase_late_bound_regions(&fn_sig);
+
+        for ((input, &ty), arg) in decl.inputs.iter().zip(fn_sig.inputs()).zip(&body.arguments) {
+            // All spans generated from a proc-macro invocation are the same...
+            if span == input.span {
+                return;
+            }
+
+            let limit = self.limit.unwrap_or_else(|| {
+                let bit_width = cx.sess().target.usize_ty.bit_width().unwrap_or(32) as u64;
+                // Cap the calculated bit width at 32-bits to reduce
+                // portability problems between 32 and 64-bit targets
+                let bit_width = cmp::min(bit_width, 32);
+                let byte_width = bit_width / 8;
+                // Use a limit of 2 times the register bit width
+                byte_width * 2
+            });
+            if_chain! {
+                if let TypeVariants::TyRef(_, ty, Mutability::MutImmutable) = ty.sty;
+                if is_copy(cx, ty);
+                if let Some(size) = cx.layout_of(ty).ok().map(|l| l.size.bytes());
+                if size <= limit;
+                if let Ty_::TyRptr(_, MutTy { ty: ref decl_ty, .. }) = input.node;
+                then {
+                    if is_self(arg) {
+                        span_lint_and_sugg(
+                            cx,
+                            TRIVIALLY_COPY_PASS_BY_REF,
+                            input.span,
+                            "this argument is passed by reference, but would be more efficient if passed by value",
+                            "consider passing by value instead",
+                            "self".into());
+                    } else {
+                        span_lint_and_sugg(
+                            cx,
+                            TRIVIALLY_COPY_PASS_BY_REF,
+                            input.span,
+                            "this argument is passed by reference, but would be more efficient if passed by value",
+                            "consider passing by value instead",
+                            snippet(cx, decl_ty.span, "_").into());
```

Comment:
The calls of the `span_lint_and_sugg` functions are exactly the same, except the `sugg` argument. Could you prettify this? Everything else lgtm.